asserts/ifacedecls_test.go

```diff
@@ -266,5 +295,231 @@ foo: [{p: "/foo/x"}, {p: "/foo/y"}]
+	// connection subrules
+	checkBoolPlugConnConstraints(c, rule.AllowConnection, false)
+	checkBoolPlugConnConstraints(c, rule.DenyConnection, true)
+	// auto-connection subrules
+	checkBoolPlugConnConstraints(c, rule.AllowAutoConnection, false)
+	checkBoolPlugConnConstraints(c, rule.DenyAutoConnection, true)
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleDefaults(c *C) {
+	rule, err := asserts.CompilePlugRule("iface", map[string]interface{}{
+		"deny-auto-connection": "true",
+	})
+	c.Assert(err, IsNil)
+
+	// everything follows the defaults...
+
+	// install subrules
+	c.Assert(rule.AllowInstallation, NotNil)
+	c.Check(rule.AllowInstallation.PlugAttributes, Equals, asserts.AlwaysMatchAttributes)
+	c.Assert(rule.DenyInstallation, NotNil)
+	c.Check(rule.DenyInstallation.PlugAttributes, Equals, asserts.NeverMatchAttributes)
+	// connection subrules
+	checkBoolPlugConnConstraints(c, rule.AllowConnection, true)
+	checkBoolPlugConnConstraints(c, rule.DenyConnection, false)
+	// auto-connection subrules
+	checkBoolPlugConnConstraints(c, rule.AllowAutoConnection, true)
+	// ... but deny-auto-connection is on
+	checkBoolPlugConnConstraints(c, rule.DenyAutoConnection, true)
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleConnectionConstraintsIDConstraints(c *C) {
+	rule, err := asserts.CompilePlugRule("iface", map[string]interface{}{
+		"allow-connection": map[string]interface{}{
+			"slot-snap-type":    []interface{}{"core", "kernel", "gadget", "app"},
+			"slot-snap-id":      []interface{}{"snapidsnapidsnapidsnapidsnapid01", "snapidsnapidsnapidsnapidsnapid02"},
+			"slot-publisher-id": []interface{}{"pubidpubidpubidpubidpubidpubid09", "canonical", "$same"},
+		},
+	})
+	c.Assert(err, IsNil)
+
+	cstrs := rule.AllowConnection
+	c.Check(cstrs.SlotSnapTypes, DeepEquals, []string{"core", "kernel", "gadget", "app"})
+	c.Check(cstrs.SlotSnapIDs, DeepEquals, []string{"snapidsnapidsnapidsnapidsnapid01", "snapidsnapidsnapidsnapidsnapid02"})
+	c.Check(cstrs.SlotPublisherIDs, DeepEquals, []string{"pubidpubidpubidpubidpubidpubid09", "canonical", "$same"})
+
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleConnectionConstraintsAttributesDefault(c *C) {
+	rule, err := asserts.CompilePlugRule("iface", map[string]interface{}{
+		"allow-connection": map[string]interface{}{
+			"slot-snap-id": []interface{}{"snapidsnapidsnapidsnapidsnapid01"},
+		},
+	})
+	c.Assert(err, IsNil)
+
+	// attributes default to always matching here
+	cstrs := rule.AllowConnection
+	c.Check(cstrs.PlugAttributes, Equals, asserts.AlwaysMatchAttributes)
+	c.Check(cstrs.SlotAttributes, Equals, asserts.AlwaysMatchAttributes)
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleErrors(c *C) {
+	tests := []struct {
+		stanza string
+		err    string
+	}{
+		{`iface: foo`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
+		{`iface:
+  - allow`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
```

Comment:
yes, you are reading this right and yes it's a generalisation I thought considering the base declaration